// @wordpress/commands/build/hooks/use-command.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = useCommand;
var _element = require("@wordpress/element");
var _data = require("@wordpress/data");
var _store = require("../store");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

/**
 * Attach a command to the command palette.
 *
 * @param {import('../store/actions').WPCommandConfig} command command config.
 */
function useCommand(command) {
  const {
    registerCommand,
    unregisterCommand
  } = (0, _data.useDispatch)(_store.store);
  const currentCallback = (0, _element.useRef)(command.callback);
  (0, _element.useEffect)(() => {
    currentCallback.current = command.callback;
  }, [command.callback]);
  (0, _element.useEffect)(() => {
    registerCommand({
      name: command.name,
      context: command.context,
      label: command.label,
      searchLabel: command.searchLabel,
      icon: command.icon,
      callback: (...args) => currentCallback.current(...args)
    });
    return () => {
      unregisterCommand(command.name);
    };
  }, [command.name, command.label, command.searchLabel, command.icon, command.context, registerCommand, unregisterCommand]);
}
